ClassicLootManager/Modules/Rosters/RosterManager/ExpressionParser.lua

```diff
@@ -0,0 +1,425 @@
+        ["+"] = function(left, right) return left + right end,
+        ["-"] = function(left, right) return left - right end
+    })
+
+    return tonumber(tokens[1])
+end
+
+local function replaceVariablesWithValues(expression, variables)
+    -- Sort the variables by length to avoid replacing substrings
+    local variableNames = {}
+    for variableName in pairs(variables) do
+        table.insert(variableNames, variableName)
+    end
+    table.sort(variableNames, function(a, b)
+        return #a > #b
+    end)
+    for _, variableName in ipairs(variableNames) do
+        local value = variables[variableName]
+        expression = expression:gsub(variableName, value)
+    end
+    return expression
+end
+
+
+-- Public methods
+ExpressionParser = {}
+function ExpressionParser:IsValidExpression(expression, variables)
+    local cleanedExpression = string.lower(expression:gsub("%s+", ""))
+    local errors = {}
+
+    if #cleanedExpression == 0 then
+        errors[#errors + 1] = CLM.L["Expression is empty."]
+        return false, errors
+    end
+
+    local failureConditions = {
+        { OPERATION_TYPE.OR, { "^[%+%/%*%^]", "^[%-][%-]" },                                        CLM.L["Expression may not start on an operator."] },
+        { OPERATION_TYPE.OR, { "[%+%-%/%*%^]$" },                                                   CLM.L["Expression may not end on an operator."] },
+        { OPERATION_TYPE.OR, { "[%+%-%/%*%^][%+%/%*%^]", "[%+%-%/%*%^][%+%-%/%*%^][%+%-%/%*%^]" },  CLM.L["Expression may not contain more than one operator in a row."] },
+        { OPERATION_TYPE.OR, { "[%+%-%/%*%^][%)]" },                                                CLM.L["Expression may not contain an operator before a closing parenthesis."] },
+        { OPERATION_TYPE.OR, { "[%(][%)]" },                                                        CLM.L["Expression may not contain empty parenthesis."] },
+        { OPERATION_TYPE.OR, { "[%d%a][%(]" },                                                      CLM.L["Expression may not contain a non-operator/non-parenthesis before an opening parenthesis."] },
+        { OPERATION_TYPE.OR, { "^%.", "[^%d]%." },                                                  CLM.L["A decimal point must be preceded by a digit."] },
+        { OPERATION_TYPE.OR, { "%.$", "%.[^%d]" },                                                  CLM.L["A decimal point must be followed by a digit."] },
+        { OPERATION_TYPE.OR, { "%d+%.%d+%." },                                                      CLM.L["A number cannot contain multiple decimal points."] }
+    }
+
+    verifyConditions(cleanedExpression, failureConditions, errors)
+    verifyAllTextIsVariableNames(cleanedExpression, variables, errors)
+    verifyParenthesisAreBalanced(cleanedExpression, errors)
+
+    return #errors == 0, errors
+end
+
+function ExpressionParser:Evaluate(expression, variables)
+    local cleanedExpression = string.lower(expression:gsub("%s+", ""))
```

Comment:
1. Shouldn't the cleaning be done once, on storing the expression?
2. Ensure no string modificators or patterns are allowed at all (`%slot_multiplier` coming from expression etc),